compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5680,9 +5680,10 @@ private BVarSymbol checkParameterNameForDefaultArgument(BLangIdentifier argName,
             }
         }
         if (incRecordParamAllowAdditionalFields != null) {
-            LinkedHashMap<String, BField> fields = ((BRecordType) incRecordParamAllowAdditionalFields.type).fields;
-            if (!fields.containsKey(argName.value) && types.isAssignable(expr.type,
-                                            ((BRecordType) incRecordParamAllowAdditionalFields.type).restFieldType)) {
+            BRecordType incRecordType = (BRecordType) incRecordParamAllowAdditionalFields.type;
+            BType exprType = checkExpr(expr, env);
```

Comment:
As discussed offline, let's check if we can use the expected type for the argument here to make use of the contextual information.

Reply:
Fixed.